### PR TITLE
Mobt 157 nbhood refactor code consolidation

### PR DESCRIPTION
**Pull request by fionaRust**
Initial neighbourhood processing refactoring. Addresses part of https://github.com/metoppv/mo-blue-team/issues/157
To be reviewed after https://github.com/metoppv/improver/pull/1648

This PR merges the key functionality from the circular kernel plugin (CircularNeighbourhood) into the SquareNeighbourhood plugin and renames the SquareNeighbourhood plugin to be simply Neighbourhood. This enables the circular kernel to take advantage of the masked data handling in the square plugin and means there is less discrepancy between the functionality of the different kernels.

This is only step 1 of the neighbourhood refactor. Follow on work includes:
* Moving the Neighbourhood plugin out of the square_kernel.py file. I didn't want to do this in this PR as I didn't want lots of moving to obscure the code changes. As part of moving the Neighbourhood plugin I'll also look at the other neighbourhood processing base classes and try and simplify them now we have a single neighbourhood plugin.
* Consolidation and checking the coverage of unit tests. This PR only ensures the existing unit tests work with the new plugin. Now the square and circular kernels share a lot of code, I will compare the unit test coverage for the square and circular functionality, move them all into one place and work out if we now have duplicate coverage in some places and not enough coverage in others. The same for acceptance tests.
* A final step  will be to look at the options in the CLIs and remove unnecessary options and/or consolidate CLIs.

### CODE REVIEW

File: improver/nbhood/nbhood.py [hunk 1]
```diff
@@ -323,10 +310,12 @@ def __init__(
             neighbourhood_method, radii, lead_times=lead_times
         )
 
-        methods = {"circular": CircularNeighbourhood, "square": SquareNeighbourhood}
```
Comment (fionaRust) [inline improver/nbhood/nbhood.py:326]: These base plugins need further tidy up, but I'm going to leave that for a separate PR to avoid complicating this one.

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 2]
```diff
@@ -88,10 +74,8 @@ def setUp(self):
     def test_basic(self):
         """Test that the plugin returns an iris.cube.Cube."""
 
-        ranges = 2
-        result = CircularNeighbourhood(weighted_mode=False).apply_circular_kernel(
```
Comment (fionaRust) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:92]: I've modified these tests so they now call the run method on Neighbourhood plugin, which testes the same functionality as before. The run method takes in the radius in metres rather than the range in grid cells.

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 3]
```diff
@@ -136,48 +117,47 @@ def test_multi_point_multitimes(self):
...
-        ranges = 3
-        result = CircularNeighbourhood(weighted_mode=True).apply_circular_kernel(
-            cube, ranges
-        )
+        radius = 6000
+        result = Neighbourhood("circular", weighted_mode=True).run(cube, radius)
         self.assertArrayAlmostEqual(result.data, expected)
 
     def test_single_point_masked_to_null(self):
-        """Test behaviour with a masked non-zero point. The behaviour here is
```
Comment (fionaRust) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:146]: Before the scipy code didn't handle masked data, so this test produced a result that was wrong. Now the masked data is handled by reusing the code from the square kernel, and this test returns a correct result. The same is true of the test below.

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 4]
```diff
@@ -200,32 +178,18 @@ def test_single_point_range_5(self):
...
-        )
+        radius = 10000
+        result = Neighbourhood("circular", weighted_mode=True).run(self.cube, radius)
         self.assertArrayAlmostEqual(result.data, expected)
 
     def test_single_point_range_5_small_domain(self):
-        """Test behaviour - non-zero point, small domain, large range.
-        This exhibits the undesirable edge reflection behaviour."""
-
+        """Test code fails if radius larger than the domain."""
```
Comment (fionaRust) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:186]: The circular kernel is now constrained to only accept radii smaller than the domain, in the same way as the square kernel

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 5]
```diff
@@ -353,37 +305,35 @@ def test_single_point_on_corner(self):
...
         self.assertArrayAlmostEqual(result.data, expected)
 
 
 class Test_run(IrisTest):
 
-    """Test the run method on the CircularNeighbourhood class."""
+    """Test the run method on the Neighbourhood class."""
 
-    RADIUS = 6100
+    RADIUS = 4000
```
Comment (fionaRust) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:317]: The tests below had to be changed slightly as the radius before was actually larger than the domain.

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 6]
```diff
@@ -396,9 +346,10 @@ def test_mask_cube(self):
         data[2, 2] = 0
         cube = set_up_variable_cube(data, spatial_grid="equalarea",)
 
-        msg = "The use of a mask cube with a circular kernel is " "not yet implemented."
```
Comment (fionaRust) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:399]: Now it is implemented so we check the result of this test

File: improver/nbhood/square_kernel.py [hunk 7]
```diff
@@ -196,13 +215,23 @@ def run(self, cube: Cube, radius: float, mask_cube: Optional[Cube] = None) -> Cu
...
         original_methods = cube.cell_methods
         grid_cells = distance_to_number_of_grid_cells(cube, radius)
         nb_size = 2 * grid_cells + 1
+        if self.neighbourhood_method == "circular":
+            axes = []
+            for axis in ["x", "y"]:
+                coord_name = cube.coord(axis=axis).name()
+                axes.append(cube.coord_dims(coord_name)[0])
+
+            self.kernel = circular_kernel(grid_cells, self.weighted_mode)
```
Comment (bayliffe) [inline improver/nbhood/square_kernel.py:233]:
```suggestion
            self.kernel = circular_kernel(grid_cells, self.weighted_mode)
        elif self.neighbourhood_method == "square":
            self.nb_size = 2 * grid_cells + 1
        else:
            raise Your Chosen Exception for an unrecognised neighbourhood_method.
```
The second elif will allow you to modify `_calculate_neighbourhood` to remove nb_size as a required argument, which it isn't if we are using a circular kernel. The else allows for an exception if the chosen neighbourhood method is not valid, as I can't see that checked in the plugin, which I think it should be.

File: improver/nbhood/nbhood.py [hunk 8]
```diff
@@ -69,7 +68,7 @@ class BaseNeighbourhoodProcessing(PostProcessingPlugin):
 
     def __init__(
         self,
-        neighbourhood_method: Union[CircularNeighbourhood, SquareNeighbourhood],
+        neighbourhood_method: Union[Neighbourhood],
```
Comment (bayliffe) [inline improver/nbhood/nbhood.py:71]: I don't think you need to pass a method in like this anymore. It is always this plugin that we are using now, just configured for either square or circular neighbourhooding. Passing in a string stating "square" or "circular" would be sufficient and remove some of the complexity around checking if this is a callable etc.

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 9]
```diff
@@ -382,9 +346,10 @@ def test_mask_cube(self):
...
-        with self.assertRaisesRegex(NotImplementedError, msg):
-            CircularNeighbourhood().run(cube, self.RADIUS, mask_cube=cube)
+        result = Neighbourhood("circular",).run(cube, self.RADIUS, mask_cube=cube)
```
Comment (bayliffe) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:349]: The doc-string for this test needs updating.

**Review by bayliffe — CHANGES_REQUESTED** [review]
No major comments in the end. Given that this is a work in progress I have tried to ignore the stuff you indicate you are still to revisit (like duplicate test coverage and the base class).

File: improver/nbhood/square_kernel.py [hunk 10]
```diff
@@ -36,21 +36,29 @@
...
+    check_if_grid_is_equal_area,
+    distance_to_number_of_grid_cells,
+)
 
 
-class SquareNeighbourhood:
+class Neighbourhood:
 
     """
     Methods for use in application of a square neighbourhood.
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver/nbhood/square_kernel.py]: Done

File: improver/nbhood/square_kernel.py [hunk 11]
```diff
@@ -196,13 +215,23 @@ def run(self, cube: Cube, radius: float, mask_cube: Optional[Cube] = None) -> Cu
...
         original_attributes = cube.attributes
         original_methods = cube.cell_methods
         grid_cells = distance_to_number_of_grid_cells(cube, radius)
         nb_size = 2 * grid_cells + 1
+        if self.neighbourhood_method == "circular":
+            axes = []
+            for axis in ["x", "y"]:
+                coord_name = cube.coord(axis=axis).name()
+                axes.append(cube.coord_dims(coord_name)[0])
+
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver/nbhood/square_kernel.py]: Done

File: improver/nbhood/square_kernel.py [hunk 12]
```diff
@@ -36,21 +36,29 @@
 import numpy as np
 from iris.cube import Cube
 from numpy import ndarray
+from scipy.ndimage.filters import correlate
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver/nbhood/square_kernel.py:39]: Done

File: improver/nbhood/square_kernel.py (hunk 7 above)
Comment (fionaRust) [reply improver/nbhood/square_kernel.py:225]: Done, I think I may check whether this is the right place for this to be raised in a future PR as part of sorting out the base class stuff.

File: improver_tests/nbhood/square_kernel/test_SquareNeighbourhood.py [hunk 13]
```diff
@@ -77,6 +77,13 @@ def setUp(self):
             data, spatial_grid="equalarea",
         )
 
+    def test_invalid_neighbourhood_method(self):
+        """Test an exception is raised if an invalid neighbourhood option
+         is provided"""
```
Comment (bayliffe) [inline improver_tests/nbhood/square_kernel/test_SquareNeighbourhood.py:82]: A single extra space preceding "is". Significant.

**Review by bayliffe — DISMISSED** [review]

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 14]
```diff
@@ -339,52 +305,50 @@ def test_single_point_on_corner(self):
...
+        self.assertIsInstance(cube, Cube)
+        self.assertArrayAlmostEqual(result.data.data, np.ones((5, 5)))
+        self.assertArrayEqual(result.data.mask, ~data.astype(np.bool))
```
Comment (gavinevans) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:351]: I think that you only need `bool` here. I think using `np.bool` is deprecated, see #1646.

File: improver/nbhood/nbhood.py (hunk 8 above)
Comment (gavinevans) [inline improver/nbhood/nbhood.py:71]: This can be just:
```suggestion
        neighbourhood_method: Neighbourhood,
```

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 15]
```diff
@@ -122,48 +117,47 @@ def test_multi_point_multitimes(self):
...
-        result = CircularNeighbourhood(weighted_mode=True).apply_circular_kernel(
-            self.cube, ranges
-        )
+        # Update expected result for rows affected by masked point
+        expected[5:9, 4:11] = [
+            [1, 0.991736, 0.966102, 0.957265, 0.966102, 0.991736, 1],
+            [1, 0.966667, 0.940171, np.nan, 0.940171, 0.966667, 1],
+            [1, 0.958677, 0.932203, 0.923077, 0.932203, 0.958677, 1],
+            [1, 0.967741, 0.942149, 0.933333, 0.942149, 0.967741, 1],
+        ]
```
Comment (gavinevans) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:156]: I think that this is overwriting all but one row within `expected` compared to lines 148-149 meaning that a combination of these two methods of setting `expected` gives the intended result. In this PR, or later, it would be good to consolidate this.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks for this consolidation, @fionaRust 👍 

I've added a few minor comments, appreciating that follow-on PRs will further refactor and tidy-up the code pathways.

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py (hunk 14 above)
Comment (fionaRust) [reply improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py]: Done

File: improver/nbhood/nbhood.py (hunk 8 above)
Comment (fionaRust) [reply improver/nbhood/nbhood.py]: Done

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py (hunk 15 above)
Comment (fionaRust) [reply improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:155]: Done

**Review by gavinevans — APPROVED** [review]
